### PR TITLE
Dev/gc refactor

### PR DESCRIPTION
**Pull request by TianlongLiang**
Modify opcodes:
1.	WASM_OP_CALL_INDIRECT
2.	WASM_OP_SELECT_T 
3.	WASM_OP_TABLE_GET 
4.	WASM_OP_TABLE_SET
5.	WASM_OP_REF_NULL 
6.	WASM_OP_REF_IS_NULL 
7.	WASM_OP_REF_FUNC
8.	WASM_OP_TABLE_INIT
9.	WASM_OP_TABLE_GROW
10.	WASM_OP_TABLE_FILL 
11.	WASM_OP_GET_LOCAL
12.	WASM_OP_SET_LOCAL
13.	WASM_OP_TEE_LOCAL

Currently, put gc APIs wrapper in a new file `aot_gc_object_wrapper.c`

### CODE REVIEW

File: core/iwasm/compilation/aot_compiler.c [hunk 1]
```diff
@@ -441,32 +441,49 @@ aot_compile_func(AOTCompContext *comp_ctx, uint32 func_index)
...
-                                               && (type_idx != VALUE_TYPE_F64)))
-                    return false;
+#if WASM_ENABLE_GC != 0
+                if (comp_ctx->enable_gc) {
+                    if (!aot_compile_op_select(
+                            comp_ctx, func_ctx,
+                            (type_idx != VALUE_TYPE_I64)
+                                && (type_idx != VALUE_TYPE_F64)
+#if UINTPTR_MAX == UINT64_MAX
+                                && (!wasm_is_type_reftype(type_idx))
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_compiler.c:465]: A little strange for the modification, WASM_ENABLE_REF_TYPES also has funcref and externref, but it didn't check whether type_idx is funcref or externref.
Seems no need to modify?

File: core/iwasm/compilation/aot_emit_variable.c [hunk 2]
```diff
@@ -17,13 +17,34 @@
...
+            case VALUE_TYPE_I64:
+            case VALUE_TYPE_F32:
+            case VALUE_TYPE_F64:
+            case VALUE_TYPE_V128:
+            case VALUE_TYPE_I1:
+                break;
+            default:
+                local_type = VALUE_TYPE_OBJECT_REF;
+        }
+    }
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_emit_variable.c:44]:
```
if (comp_ctx->enable_gc && wasm_is_type_reftype(local_type))
    local_type = VALUE_TYPE_GC_REF;
```

File: core/iwasm/compilation/aot_emit_table.c [hunk 3]
```diff
@@ -293,10 +342,18 @@ aot_compile_op_table_init(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
     param_types[5] = I32_TYPE;
     ret_type = VOID_TYPE;
 
-    if (comp_ctx->is_jit_mode)
-        GET_AOT_FUNCTION(llvm_jit_table_init, 6);
+#if WASM_ENABLE_GC != 0
+    if (comp_ctx->enable_gc) {
+        GET_AOT_FUNCTION(aot_gc_table_init, 6);
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_emit_table.c:347]: Just use the original function, don't change here.

File: core/iwasm/compilation/aot_emit_table.c [hunk 4]
```diff
@@ -428,7 +485,15 @@ aot_compile_op_table_grow(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
     param_types[0] = INT8_PTR_TYPE;
     param_types[1] = I32_TYPE;
     param_types[2] = I32_TYPE;
-    param_types[3] = I32_TYPE;
+#if WASM_ENABLE_GC != 0
+    if (comp_ctx->enable_gc) {
+        param_types[3] = OBJECT_REF_TYPE;
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_emit_table.c:490]: No need to change for table grow?

File: core/iwasm/compilation/gc/aot_gc_object_wrapper.c [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                                                    func_ctx->func, name))) { \
+            aot_set_last_error("llvm add basic block failed.");               \
+            goto fail;                                                        \
+        }                                                                     \
+    } while (0)
+
+bool
+aot_call_wasm_create_func_obj(AOTCompContext *comp_ctx,
+                              AOTFuncContext *func_ctx, LLVMValueRef func_idx,
+                              LLVMValueRef *p_gc_obj)
```
[the remark this answers is not in the document]
Comment (TianlongLiang) [reply core/iwasm/compilation/gc/aot_gc_object_wrapper.c:26]: It's also used for the new opcode `ref.func`, so I put it in a separate file. 
I currently put all wrappers in `aot_gc_object_wrapper.c` and the new opcode related functions like `aot_compile_op_struct_new` in `aot_emit_reference.c`. 
Should I put all of them in one file `aot_emit_reference.c`? In that case, won't need to modify the cmake config files

File: core/iwasm/aot/aot_runtime.c [hunk 6]
```diff
@@ -2535,6 +2535,64 @@ aot_table_init(AOTModuleInstance *module_inst, uint32 tbl_idx,
                 length * sizeof(table_elem_type_t));
 }
 
+#if WASM_ENABLE_GC != 0
+void
+aot_gc_table_init(AOTModuleInstance *module_inst, uint32 tbl_idx,
```
[the remark this answers is not in the document]
Comment (TianlongLiang) [reply core/iwasm/aot/aot_runtime.c:2540]: I was thinking of a corner case:

The enable_gc is disabled in wamrc, it's only generates an old version aot file containing table.init opcode; but the WASM_ENALBE_GC is enabled in iwasm, when running `iwasm test.aot`, it calls the new version of aot_table_init(since macro is enabled), it will cause a problem.

So I used a different function to distinguish which function needs to be called.

File: core/iwasm/compilation/aot_compiler.c (hunk 1 above)
Comment (TianlongLiang) [reply core/iwasm/compilation/aot_compiler.c:465]: The previous select_t doesn't take the reference as an argument.

File: core/iwasm/compilation/aot_emit_table.c (hunk 4 above)
Comment (TianlongLiang) [reply core/iwasm/compilation/aot_emit_table.c:490]: I just tried to make the function parameter's type match with the `aot_table_grow`. When GC is enabled, the table_elem_type_t is void* rather than unsigned long int. 

When on a 64-bit platform, is it possible to become a problem if don't change it?

File: core/iwasm/compilation/gc/aot_gc_object_wrapper.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/compilation/gc/aot_gc_object_wrapper.c]: Seems it should be table.init calling aot_create_func_obj but not struct.new.
It is OK to put it in aot_emit_reference.c.

File: core/iwasm/compilation/aot_compiler.c (hunk 1 above)
Comment (wenyongh) [reply core/iwasm/compilation/aot_compiler.c:469]:
```C
                                && (type_idx != VALUE_TYPE_I64)
                                && (type_idx != VALUE_TYPE_F64)
#if UINTPTR_MAX == UINT64_MAX
#if WASM_ENABLE_GC != 0
                                && !(comp_ctx->enable_gc && wasm_is_type_reftype(type_idx))
#endif
#endfi

File: core/iwasm/compilation/aot_emit_gc.c [hunk 7]
```diff
@@ -0,0 +1,80 @@
+#include "aot_emit_gc.h"
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_emit_gc.c:1]: Should add license header.

File: core/iwasm/interpreter/wasm_runtime.c [hunk 8]
```diff
@@ -3809,6 +3809,53 @@ llvm_jit_free_frame(WASMExecEnv *exec_env)
...
+    WASMFuncObjectRef func_obj;
+    WASMFunctionInstance *func_inst;
+    WASMFuncType *func_type;
+    uint32 type_idx;
+
+#if WASM_ENABLE_JIT != 0
+    if (Wasm_Module_AoT == module_inst->module_type) {
+        return aot_create_func_obj(module_inst, func_idx);
+    }
+#endif
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:3827]: Just `bh_assert(module_inst->module_type == Wasm_Module_Bytecode)`, the similar codes in this file have been modified in main branch.

File: core/iwasm/aot/aot_runtime.c [hunk 9]
```diff
@@ -2501,6 +2501,12 @@ aot_table_init(AOTModuleInstance *module_inst, uint32 tbl_idx,
     AOTTableInstance *tbl_inst;
     AOTTableInitData *tbl_seg;
     const AOTModule *module = (AOTModule *)module_inst->module;
+#if WASM_ENABLE_GC != 0
+    table_elem_type_t **table_elems;
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:2505]: Shold be `table_elem_type_t *table_elems`, table_elem_type_t is `void *` or `uintptr_t` when GC is enabled or not.

File: core/iwasm/aot/aot_runtime.c [hunk 10]
```diff
@@ -2528,11 +2534,31 @@ aot_table_init(AOTModuleInstance *module_inst, uint32 tbl_idx,
         return;
     }
 
+#if WASM_ENABLE_GC != 0
+    table_elems = (table_elem_type_t **)(tbl_inst->elems + dst_offset);
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:2538]: Can change to: `table_elems = tbl_inst->elems + dst_offset`

File: core/iwasm/aot/aot_runtime.c [hunk 11]
```diff
@@ -3385,3 +3411,51 @@ aot_dump_pgo_prof_data_to_buf(AOTModuleInstance *module_inst, char *buf,
...
+    WASMFunctionInstance *func_inst;
+    WASMFuncType *func_type;
+    uint32 type_idx;
+
+    if (throw_exce) {
+        error_buf = module_inst->cur_exception;
+        error_buf_size = sizeof(module_inst->cur_exception);
+    }
+
+    if (func_idx >= module_inst->e->function_count) {
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:3433]: module_inst is type AOTModuleInstance, `module_inst->e` is type AOTModuleInstanceExtra but not WASMModuleInstanceExtra, AOTModuleInstanceExtra hasn't field function_count.

Use `func_idx >= module->import_func_count + module->func_count` instead

File: core/iwasm/aot/aot_runtime.h [hunk 12]
```diff
@@ -157,6 +157,8 @@ typedef struct AOTModule {
     uint32 mem_init_data_count;
     AOTMemInitData **mem_init_data_list;
 
+    /* TODO: for array init data */
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.h:160]: What array init data means?

File: core/iwasm/compilation/aot_emit_gc.h [hunk 13]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (C) 2019 Intel Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#ifndef _AOT_EMIT_GC_H_
+#define _AOT_EMIT_GC_H_
+
+#include "aot_compiler.h"
+#include "stdbool.h"
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_emit_gc.h:10]: No need to include stdbool.h? It should have been included in platform_internal.h, or bool type should have been defined.

File: core/iwasm/aot/aot_runtime.c [hunk 14]
```diff
@@ -3385,3 +3411,51 @@ aot_dump_pgo_prof_data_to_buf(AOTModuleInstance *module_inst, char *buf,
...
+        set_error_buf_v(error_buf, error_buf_size, "unknown function %d",
+                        func_idx);
+        return NULL;
+    }
+
+    /* TODO: need obtain function instance, maybe we need to create function
+     * instances for all func? */
+    func_type = func_inst->is_import_func ? func_inst->u.func_import->func_type
+                                          : func_inst->u.func.func_type;
+    type_idx = module->func_type_indexes[func_idx];
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:3443]: Seems we can just change as:
```C
type_idx = module_inst->func_type_indexes[func_idx];
func_type = module->types[type_idx];
```

File: core/iwasm/compilation/aot_compiler.c [hunk 15]
```diff
@@ -361,13 +361,17 @@ aot_compile_func(AOTCompContext *comp_ctx, uint32 func_index)
 
                 read_leb_uint32(frame_ip, frame_ip_end, type_idx);
 
+                if (false
+#if WASM_ENABLE_GC != 0
+                    || comp_ctx->enable_gc
```
Comment (no1wudi) [inline core/iwasm/compilation/aot_compiler.c:366]: enable_gc is always available in AOT compiler, no need to guard it with WASM_ENABLE_GC

File: core/iwasm/compilation/aot_compiler.c [hunk 16]
```diff
@@ -361,13 +361,17 @@ aot_compile_func(AOTCompContext *comp_ctx, uint32 func_index)
...
+#endif
 #if WASM_ENABLE_REF_TYPES != 0
-                if (comp_ctx->enable_ref_types) {
+                    || comp_ctx->enable_ref_types
```
Comment (no1wudi) [inline core/iwasm/compilation/aot_compiler.c:369]: Ditto

File: core/iwasm/compilation/aot_compiler.c [hunk 17]
```diff
@@ -401,13 +405,17 @@ aot_compile_func(AOTCompContext *comp_ctx, uint32 func_index)
                 }
 
                 read_leb_uint32(frame_ip, frame_ip_end, type_idx);
+                if (false
+#if WASM_ENABLE_GC != 0
+                    || comp_ctx->enable_gc
```
Comment (no1wudi) [inline core/iwasm/compilation/aot_compiler.c:410]: Ditto

File: core/iwasm/compilation/aot_compiler.c [hunk 18]
```diff
@@ -401,13 +405,17 @@ aot_compile_func(AOTCompContext *comp_ctx, uint32 func_index)
...
+#endif
 #if WASM_ENABLE_REF_TYPES != 0
-                if (comp_ctx->enable_ref_types) {
+                    || comp_ctx->enable_ref_types
```
Comment (no1wudi) [inline core/iwasm/compilation/aot_compiler.c:413]: Ditto

File: core/iwasm/compilation/aot_emit_function.c [hunk 19]
```diff
@@ -1557,7 +1645,15 @@ aot_compile_op_call_indirect(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
 bool
 aot_compile_op_ref_null(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx)
 {
-    PUSH_I32(REF_NULL);
+#if WASM_ENABLE_GC != 0
+    if (comp_ctx->enable_gc) {
+        PUSH_REF(GC_REF_NULL);
```
Comment (no1wudi) [inline core/iwasm/compilation/aot_emit_function.c:1650]: How about define PUSH_REF and POP_REF as empty macro if GC is disabled, to reduce the guarded code blocks

Comment (TianlongLiang) [reply core/iwasm/compilation/aot_emit_function.c:1641]: Now only use the macro for necessary blocks

File: core/iwasm/compilation/aot_compiler.c [hunk 20]
```diff
@@ -361,13 +361,11 @@ aot_compile_func(AOTCompContext *comp_ctx, uint32 func_index)
 
                 read_leb_uint32(frame_ip, frame_ip_end, type_idx);
 
-#if WASM_ENABLE_REF_TYPES != 0
-                if (comp_ctx->enable_ref_types) {
+                if (false || comp_ctx->enable_gc
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_compiler.c:364]: No need to add `false ||`

File: core/iwasm/compilation/aot_compiler.c [hunk 21]
```diff
@@ -401,13 +399,11 @@ aot_compile_func(AOTCompContext *comp_ctx, uint32 func_index)
                 }
 
                 read_leb_uint32(frame_ip, frame_ip_end, type_idx);
-#if WASM_ENABLE_REF_TYPES != 0
-                if (comp_ctx->enable_ref_types) {
+                if (false || comp_ctx->enable_gc
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_compiler.c:402]: Same as above

File: core/iwasm/compilation/aot_compiler.h [hunk 22]
```diff
@@ -163,7 +163,15 @@ check_type_compatible(uint8 src_type, uint8 dst_type)
 #define POP_V128(v) POP(v, VALUE_TYPE_V128)
 #define POP_FUNCREF(v) POP(v, VALUE_TYPE_FUNCREF)
 #define POP_EXTERNREF(v) POP(v, VALUE_TYPE_EXTERNREF)
-
+#if WASM_ENABLE_GC != 0
+#define POP_REF(v) POP(v, VALUE_TYPE_GC_REF)
+#else
+#define POP_REF(v)                                                        \
+    do {                                                                  \
+        LOG_DEBUG("enable_gc=true when CMake flag are not set properly"); \
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_compiler.h:171]: How about `bh_assert(!"should not POP_REF when WASM_ENABLE_GC macro isn't enabled");`

File: core/iwasm/compilation/aot_compiler.h [hunk 23]
```diff
@@ -217,6 +225,15 @@ check_type_compatible(uint8 src_type, uint8 dst_type)
 #define PUSH_COND(v) PUSH(v, VALUE_TYPE_I1)
 #define PUSH_FUNCREF(v) PUSH(v, VALUE_TYPE_FUNCREF)
 #define PUSH_EXTERNREF(v) PUSH(v, VALUE_TYPE_EXTERNREF)
+#if WASM_ENABLE_GC != 0
+#define PUSH_REF(v) PUSH(v, VALUE_TYPE_GC_REF)
+#else
+#define PUSH_REF(v)                                                       \
+    do {                                                                  \
+        LOG_DEBUG("enable_gc=true when CMake flag are not set properly"); \
+        assert(0);                                                        \
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_compiler.h:234]: How about bh_assert(!"should not PUSH_REF when WASM_ENABLE_GC macro isn't enabled");

**Review by no1wudi — APPROVED** [review]
LGTM
